src/estimators/same_key_aggregator.py

```diff
@@ -94,54 +98,107 @@ def assert_compatible(self, other):
+        for candidate classes.
+      frequency_noiser_class: a class of noise to be added to active_key_histogram.
+        See wfa_cardinality_estimation_evaluation_framework.estimators.estimator_noisers
+        for candidate classes.
+      reach_epsilon: DP-epsilon for the noise on one_plus_reach.
+      frequency_epsilon: DP-epsilon for the noise on active_key_histogram.
+      reach_delta: DP-delta for the noise on one_plus_reach. Set as 0 if
+        reach_noise_class guarantees purely epsilon-DP.
+      frequency_delta: DP-delta for the noise on active_key_histogram. Set as 0
+        if frequency_noise_class guarantees purely epsilon-DP.
+      reach_noiser_kwargs: a dictionary of other kwargs to be specified in
+        reach_noiser_class. For example, it can be an empty dictionary, or
+        {'random_state': <a np.random.RandomState>}.
+      frequency_noiser_kwargs: a dictionary of other kwargs to be specified in
+        frequency_noiser_class. For example, it can be an empty dictionary, or
+        {'random_state': <a np.random.RandomState>}.
     """
     self.max_freq = max_freq
-    self.one_plus_reach_noiser = None
-    self.histogram_noiser = None
-    if noiser_class is not None:
-      assert epsilon_split > 0 and epsilon_split < 1, (
-          'In StandardizedHistogramEstimator, epsilon_split must be >0 and <1.')
-      # self.one_plus_reach_noiser is the noiser on the estimate of 1+ reach.
-      self.one_plus_reach_noiser = noiser_class(epsilon=epsilon * epsilon_split)
-      # self.histogram_noiser is the noiser on the frequency histogram among
-      # effective keys
-      # For simplicity, suppose the two noisers share the same noiser_class.
-      self.histogram_noiser = noiser_class(
-          epsilon=epsilon * (1 - epsilon_split))
+    self.one_plus_reach_noiser, self.reach_epsilon, self.reach_delta = (
+        StandardizedHistogramEstimator.define_noiser(
+            reach_noiser_class, reach_epsilon, reach_delta, reach_noiser_kwargs))
+    self.histogram_noiser, self.frequency_epsilon, self.frequency_delta = (
+        StandardizedHistogramEstimator.define_noiser(
+            frequency_noiser_class, frequency_epsilon,
+            frequency_delta, frequency_noiser_kwargs))
+    self.total_epsilon, self.total_delta = self.calculate_total_privacy_budget()
 
   def __call__(self, sketch_list):
     ska = StandardizedHistogramEstimator.merge_sketch_list(sketch_list)
     return self.estimate_cardinality(ska)
 
+  @classmethod
+  def define_noiser(cls, noiser_class, epsilon, delta=0, kwargs={}):
+    """A util for defining reach_noiser and frequency_noiser.
+
+    Args:
+      noiser_class: given name of noiser_class.
+      epsilon: given epsilon for the noiser.
+      delta: given delta for the noiser.
+      kwargs: other args to be specified in the noiser_class.
+
+    Returns:
+      a tuple of (<obtained noiser>, <actual epsilon the noiser guarantees>,
```

Comment:
What is an example when this is smaller than the input epsilon?
(I can imagine the delta case where delta = 0 for Laplace or something like that but not surer about epsilon.)